event-loop.c

```diff
@@ -10,15 +10,77 @@
 
 #include "kanshi.h"
 
+static int set_pipe_flags(int fd) {
+	int flags = fcntl(fd, F_GETFL);
+	if (flags == -1) {
+		fprintf(stderr, "fnctl F_GETFL failed: %s\n", strerror(errno));
+		return -1;
+	}
+	flags |= O_NONBLOCK;
+	if (fcntl(fd, F_SETFL, flags) == -1) {
+		fprintf(stderr, "fnctl F_SETFL failed: %s\n", strerror(errno));
+		return -1;
+	}
+	flags = fcntl(fd, F_GETFD);
+	if (flags == -1) {
+		fprintf(stderr, "fnctl F_GETFD failed: %s\n", strerror(errno));
+		return -1;
+	}
+	flags |= O_CLOEXEC;
+	if (fcntl(fd, F_SETFD, flags) == -1) {
+		fprintf(stderr, "fnctl F_SETFD failed: %s\n", strerror(errno));
+		return -1;
+	}
+	return 0;
+}
+
+static int signal_pipefds[2];
+
+static void signal_handler(int signum) {
+	if (write(signal_pipefds[1], &signum, sizeof(signum)) == -1) {
+		// we can't use stdio inside a signal handler
```

Comment:
TBH, we can just call `abort()` here and stop trying to do anything more.

Reply:
great, that simplifies things a bit :)